llvm/lib/TableGen/Record.cpp

```diff
@@ -987,6 +987,29 @@ Init *UnOpInit::Fold(Record *CurRec, bool IsFinal) const {
       }
     }
     break;
+
+  case LISTFLATTEN:
+    if (ListInit *LHSList = dyn_cast<ListInit>(LHS)) {
+      ListRecTy *InnerListTy = dyn_cast<ListRecTy>(LHSList->getElementType());
+      // list of non-lists, !listflatten() is a NOP.
+      if (!InnerListTy)
+        return LHS;
+      std::vector<Init *> Flattened;
+      auto CanFlatten = [LHSList, &Flattened]() {
+        // Concatenate elements of all the inner lists.
+        for (Init *InnerInit : LHSList->getValues()) {
+          ListInit *InnerList = dyn_cast<ListInit>(InnerInit);
+          if (!InnerList)
+            return false;
+          for (Init *InnerElem : InnerList->getValues())
+            Flattened.push_back(InnerElem);
+        };
+        return true;
+      }();
```

Comment:
No (), this is producing a no-op call?

Reply:
It would be less surprising to move the call down to if (CanFlatten()) than have auto as bool

Reply:
I changed this to have a non-side effecting lambda that returns an std::optional<std::vector<Init *>>. PTAL.